keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,231 @@
+downloaded and the estimated pull time based on the currently available
+bandwidth. These information are directly available via tools like `kubectl`,
+because each workload exposes this state directly to the Kubernetes API.
+
+## Motivation
+
+During Kubernetes workload creation, the time needed to pull the container image
+from the remote location varies depending on multiple factors. In low-bandwidth
+scenarios the image pull step might consume a significant amount time
+especially if the container image is large. Kubernetes provides currently no
+possibility to expose this pull progress to the user, whereas the only indicator
+is the `ContainerCreating` or `PullingImage` state.
+
+### Goals
+
+- Enhance the Container Runtime Interface (CRI) to be able to stream all
+  necessary information from the underlying container runtime
+- Expose the progress to the Kubernetes API
+- Providing a way how to display the progress to the end user via the CLI
+
+### Non-Goals
+
+- Everything which is not related to the container image pull procedure
+
+## Proposal / User Stories
+
+The overall implementation can be split up into three user stories.
+
+### CRI API enhancement
+
+The Kubernets runtime API `ImageService` has to be modified to provide an
+additional streaming RPC. This server-side stream provides continuously image
+pull related metadata during the whole image pull process. The new endpoint
+would be added in addition to the already available `PullImage` RPC and re-use
+the request data type:
+
+```protobuf
+service ImageService {
+    …
+
+    // PullImageStream pulls an image and provides continuous metadata during
+    // that time
+    rpc PullImageProgress(PullImageRequest) returns (stream PullImageProgressResponse) {}
+}
+
+enum PullImageState {
+    STARTED = 0;
+    PULLING = 1;
+    FAILED  = 2;
+    DONE    = 3;
+}
+
+message PullImageProgressResponse {
+    // The current state of the image pull proress
+    PullImageState state = 1;
+
+    // The overall size of the image in bytes
+    uint64 size = 2;
+
+    // The amount of data already retrieved in bytes
+    uint64 current_offset = 3;
+
+    // String indicating the reason for the current state, like a failure message
+    string reason = 4;
+}
+```
+
+The `PullImageState` provides richer information to the kubelet in which state
+the image pull currently resides as well as a fine granular error messages
+directly from the runtimes via the `reason` field.
+
+### Kubernetes API additions
+
+The already existing kubelet `imageManager` events `PullingImage`, `PulledImage`
+and `FailedToPullImage` will be re-used and extended with the additional
+metadata provided by the CRI. This means that the container will still reside in
+`ContainerStateWaiting` during the image pull, but the events regarding the
+image pull process will be thrown periodically. The amount of how often the
+kubelet throws this event should be configurable in a range between 1s and 1min.
```

Comment:
BTW - I'm not 100% convinced about the use of Events type for it. 
I would like to see a bit more thoughts about what will it mean once we will migrate everywhere to new events API:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/events-redesign.md
https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190131-event-series.md

[Note that we already migrated scheduler in 1.16, and we're planning to do the remaining components, including kubelet in 1.18]

Reply:
Sounds reasonable, maybe @kubernetes/sig-node-api-reviews could provide more input here?

Reply:
> the events regarding the image pull process will be thrown periodically

What's the period time? 30s? 1min?
Or the progress for every 10% will be reported?

This will definitely become a heavy burden when the cluster scales out. Thinking about deploying a DaemonSet.

Reply:
> So to make my point a bit more explicit - I would like to see some discussion whether we should use Events or some dedicated CRD for this maybe. Both has advantages and disadvantages and the answer isn't clear for me.

Can you please point out the advantages of using a CRD over an Event in this case? This is not clear to me in all shapes.

> What's the period time? 30s? 1min?

The period is the configured time. :)

> Or the progress for every 10% will be reported?

Hm I don't think that this will increase the usability for the user if a large image only reports at every 10%. I think a time-constraint makes most sense, WDYT?

> This will definitely become a heavy burden when the cluster scales out. Thinking about deploying a DaemonSet.

Do you think that we should add a global limit per image/node?

Reply:
> Can you please point out the advantages of using a CRD over an Event in this case? This is not clear to me in all shapes.

Event is already defined API.
- Do we really want to change hide the number in the event message?
- Events have TTL=1h - do we really want it? Maybe CR with OwnerReferences set to the pod is better?
- Events are being merged (though we need to check based on what). What exactly do we want? Having a single CR per pod that is being explictly updated (rather than `emited` as events are)?
- If we use a dedicated type, I can imagine dedicated client, with explicit QPS limits. Medium/long term, a dedicated bucket for request priority/fairness quota.

Just some advantages that I can see. but there are also disadvantages (like dependency on the new CRD, how this is going to be installed, additional maintainence burden, ...).

Reply:
i think we need to figure out a budget on the number of writes back to an api server we anticipate during pod startup.  i see this as potentially overwhelming the api server in dense clusters.

Reply:
Yeah I agree. So after the KubeCon and some discussions around the topic API vs CRD I tend to go the more future proof CRD way. Would it be okay to just state:

> The image pull progress feature will be deployed via a Custom Resource Definition (CRD),
> which will be deployed automatically via a new feature gate `ImagePullProgress`.

WDYT? If we can agree on that I'm going to update the PR.

Reply:
So is there now support for deploying a CRD that core Kubernetes can depend on? Do you have a pointer to an example or documentation for how this "will be deployed automatically via a new feature gate" part can be implemented? I'm curious because I am facing the same question in another KEP. Thanks!

Reply:
> So is there now support for deploying a CRD that core Kubernetes can depend on?

I think that I don't fully agree on the fact that it will "depend on it". Basically, this is more information surfaced for debugging (the fact that we consider putting it in events, which are not guaranteed to be delivered, is also a good sign of it). So I guess what I'm saying here, is that it's not a critical feature - it's more "you will have less information for debugging why things are taking so long".

Reply:
@wojtek-t do you think that we can ship the CRD with the kubelet (feature gate)? Or do you think it should live outside of the k/k tree?

Reply:
I don't know - we should discuss this topic on SIG Architecture. Next meeting will happen next week - we should probably add this topic to agenda.

Reply:
That would be wonderful, is the meeting somehow timezone friendly to people from Europe? :)

Reply:
11am PST - so not perfect, but attendable

Reply:
@wojtek-t :
> I think that I don't fully agree on the fact that it will "depend on it".

Let's not rathole on how important the CRD is for Kubernetes.

The key question is whether any code in the kubernetes/kubernetes repo will depend on the client bindings for the CRD (because that affects where to maintain, generate and version them) and who's going to install the CRD when bringing up a cluster. If that part is left to tools like kubeadm, then Kubernetes itself has to work also when the CRD is not installed.

@saschagrunert:
> That would be wonderful, is the meeting somehow timezone friendly to people from Europe? :)

https://github.com/kubernetes/community/tree/master/sig-architecture

Okayish I guess, but I know what you mean... :grimacing:

Reply:
Okay cool, I will attend :)